File: PakingLotDemo/src/model/ParkingSpot.java
package model;

import enums.VehicleSize;

import java.util.concurrent.atomic.AtomicBoolean;

/**
 * A single parking space with a fixed size. Can be occupied by at most one vehicle
 * of matching size. Thread-safe: {@link #parkVehicle(Vehicle)} uses synchronization
 * to prevent two vehicles from being assigned to the same spot.
 */
public class ParkingSpot {

    private final String spotId;
    private final VehicleSize spotSize;
    private final AtomicBoolean isOccupied;
    private Vehicle parkedVehicle;

    public boolean isOccupied() {
        return isOccupied.get();
    }

    public VehicleSize getSpotSize() {
        return spotSize;
    }

    public String getSpotId() {
        return spotId;
    }

    public ParkingSpot(VehicleSize size, String id) {
        this.spotSize = size;
        this.spotId = id;
        this.isOccupied = new AtomicBoolean(false);
    }

    /**
     * Assigns the vehicle to this spot and returns a ticket. Thread-safe (synchronized).
     * Typically called after a strategy returns this spot from findSpot (which uses
     * {@link #canFitVehicle(Vehicle)}). Returns null if the spot was taken by another
     * thread or the vehicle size does not match this spot.
     *
     * @param vehicle the vehicle to park
     * @return the parking ticket linking this spot and vehicle, or null if spot is occupied or size does not fit
     */
    public synchronized ParkingTicket parkVehicle(Vehicle vehicle) {
        if (isOccupied.get() || spotSize != vehicle.getVehicleSize())
            return null;
        isOccupied.set(true);
        parkedVehicle = vehicle;
        return new ParkingTicket(vehicle, this);
    }

    /**
     * Releases this spot so it can be used again. Called only after successful payment at exit.
     */
    public void unParkVehicle() {
        isOccupied.set(false);
        parkedVehicle = null;
    }

    /**
     * Returns true if this spot fits the vehicle and is free. Used by strategies to find candidates.
     * Does not claim the spot; {@link #parkVehicle(Vehicle)} does that under lock and may return null if taken.
     *
     * @param vehicle the vehicle to check
     * @return true if the spot is free and size matches, false otherwise
     */
    public boolean canFitVehicle(Vehicle vehicle) {
        return !isOccupied.get() && spotSize == vehicle.getVehicleSize();
    }
}

File: problems/python/ParkingLotSystem/src/model/parking_spot.py
"""A single parking space."""
from __future__ import annotations
import threading
from typing import Optional
from enums.vehicle_size import VehicleSize
from model.vehicle import Vehicle
from model.parking_ticket import ParkingTicket


class ParkingSpot:
    def __init__(self, size: VehicleSize, spot_id: str):
        self._spot_id = spot_id
        self._spot_size = size
        self._is_occupied = False
        self._lock = threading.Lock()
        self._parked_vehicle: Optional[Vehicle] = None

    def is_occupied(self) -> bool:
        return self._is_occupied

    def get_spot_size(self) -> VehicleSize:
        return self._spot_size

    def get_spot_id(self) -> str:
        return self._spot_id

    def can_fit_vehicle(self, vehicle: Vehicle) -> bool:
        return not self._is_occupied and self._spot_size == vehicle.get_vehicle_size()

    def park_vehicle(self, vehicle: Vehicle) -> Optional[ParkingTicket]:
        with self._lock:
            if self._is_occupied or self._spot_size != vehicle.get_vehicle_size():
                return None
            self._is_occupied = True
            self._parked_vehicle = vehicle
            return ParkingTicket(vehicle, self)

    def un_park_vehicle(self) -> None:
        self._is_occupied = False
        self._parked_vehicle = None
